feat: Expose pii config validation over cabi

untitaker opened the pull request.

untitaker committed:
feat: Expose pii config validation over cabi

jan-auer commented on py/sentry_relay/processing.py:
```diff
@@ -165,6 +167,21 @@ def is_glob_match(
+    try:
+        rustcall(lib.relay_validate_pii_config, encode_str(config))
+    except SerdeJsonError as e:
+        raise ValueError(e.message.split("\n")[0])
```
Can’t you just throw the original error through?

untitaker replied:
I would have to split off the useless stacktrace there then

jan-auer replied:
I see. In symbolic we do this in the abstraction already. 

It's a little unfortunate that we need to serialize a backtrace and don't just return the message. Yet, we can't return an option nicely via FFI. I'm wondering how you'd call this function from JS in the end?

untitaker replied:
For now I am only calling it in the API endpoint when persisting, there's already some nice API there to make sure this ends up in the frontend ok

mitsuhiko replied:
I think there is a potential unicode hazard in there if the message contains data outside ASCII.

jan-auer commented on relay-cabi/src/processing.rs:
```diff
@@ -172,6 +172,15 @@ ffi_fn! {
 }
 
+ffi_fn! {
+    unsafe fn relay_validate_pii_config(
```
Nit: doc comment :)

untitaker replied:
Done. I am not sure if ffi_fn will honor it

jan-auer replied:
It does, see other functions :) The cool thing is that docs are then also in the generated header.

jan-auer commented on relay-cabi/src/core.rs:
```diff
@@ -60,6 +63,9 @@ impl RelayErrorCode {
                     _ => RelayErrorCode::Unknown,
                 };
             }
+            if let Some(..) = cause.downcast_ref::<serde_json::Error>() {
```
Nit: Use `is_some` instead?

untitaker replied:
This is more consistent with the surrounding code

untitaker committed:
ref: Add docstring

untitaker committed:
fix: Fix tests

untitaker committed:
ref: Do not raise rust error

After some feedback we decided not to use errors for flow control in
Rust. The reason is mostly performance of error handling and unnecessary
backtrace we need to rip out.

jan-auer reviewed (approved):
Looking good. What Armin says still applies, but if you say it's fine I'm okay with accepting the risk.
